cli/tools/bench.rs

```diff
@@ -0,0 +1,541 @@
+}
+
+/// Type check a collection of module and document specifiers.
+async fn check_specifiers(
+  ps: &ProcState,
+  permissions: Permissions,
+  specifiers: Vec<ModuleSpecifier>,
+  lib: emit::TypeLib,
+) -> Result<(), AnyError> {
+  let module_specifiers = specifiers
+    .into_iter()
+    .map(|s| (s, ModuleKind::Esm))
+    .collect();
+  ps.prepare_module_load(
+    module_specifiers,
+    false,
+    lib,
+    Permissions::allow_all(),
+    permissions,
+    true,
+  )
+  .await?;
+
+  Ok(())
+}
+
+/// Test a single specifier as documentation containing test programs, an executable test module or
+/// both.
+async fn bench_specifier(
+  ps: ProcState,
+  permissions: Permissions,
+  specifier: ModuleSpecifier,
+  channel: Sender<BenchEvent>,
+  options: BenchSpecifierOptions,
+) -> Result<(), AnyError> {
+  let mut worker = create_main_worker(
+    &ps,
+    specifier.clone(),
+    permissions,
+    vec![ops::bench::init(channel.clone())],
+  );
+
+  // Enable op call tracing in core to enable better debugging of op sanitizer
+  // failures.
+  worker
+    .execute_script(&located_script_name!(), "Deno.core.enableOpCallTracing();")
+    .unwrap();
+
+  if options.compat_mode {
+    worker.execute_side_module(&compat::GLOBAL_URL).await?;
+    worker.execute_side_module(&compat::MODULE_URL).await?;
+
+    let use_esm_loader = compat::check_if_should_use_esm_loader(&specifier)?;
+
+    if use_esm_loader {
+      worker.execute_side_module(&specifier).await?;
+    } else {
+      compat::load_cjs_module(
+        &mut worker.js_runtime,
+        &specifier.to_file_path().unwrap().display().to_string(),
+        false,
+      )?;
+      worker.run_event_loop(false).await?;
+    }
+  } else {
+    // We execute the module module as a side module so that import.meta.main is not set.
+    worker.execute_side_module(&specifier).await?;
+  }
+
+  worker.dispatch_load_event(&located_script_name!())?;
+
+  let test_result = worker.js_runtime.execute_script(
+    &located_script_name!(),
+    &format!(
+      r#"Deno[Deno.internal].runBenchmarks({})"#,
+      json!({
+        "filter": options.filter,
+      }),
+    ),
+  )?;
+
+  worker.js_runtime.resolve_value(test_result).await?;
+
+  worker.dispatch_unload_event(&located_script_name!())?;
+
+  Ok(())
+}
+
+/// Test a collection of specifiers with test modes concurrently.
+async fn bench_specifiers(
+  ps: ProcState,
+  permissions: Permissions,
+  specifiers: Vec<ModuleSpecifier>,
+  options: BenchSpecifierOptions,
+) -> Result<(), AnyError> {
+  let log_level = ps.flags.log_level;
+
+  let (sender, receiver) = channel::<BenchEvent>();
+
+  let join_handles = specifiers.iter().map(move |specifier| {
+    let ps = ps.clone();
+    let permissions = permissions.clone();
+    let specifier = specifier.clone();
+    let sender = sender.clone();
+    let options = options.clone();
+
+    tokio::task::spawn_blocking(move || {
+      let join_handle = std::thread::spawn(move || {
```

Comment:
I've always wondered this (IIRC I've seen this same pattern elsewhere), but why do we spawn another thread here when we already have a thread from tokio? I'm guessing this is setting some thread local state?

If so, I don't believe we need to take away a thread from tokio (spawn_blocking) to do this and could instead just work with `std::thread::spawn` directly.

Reply:
That's a good point, I think this could be removed

Reply:
Opened https://github.com/denoland/deno/pull/13877